### PR TITLE
client: add NetworkStatus to Allocation

### PR DESCRIPTION
**Pull request by nickethier**
This PR adds a new field `NetworkStatus` to the `Allocation` struct which is populated by the client. With network namespaces, before an allocation starts we do not know the IP address inside the namespace. This new field allows for this IP address to be discoverable in the Nomad API. The implementation reuses the existing machinery to push client state to the servers.

A future PR will implement using this address in service registrations.

### CODE REVIEW

**Comment by notnoop** [comment]
Mind if you provide a PR summary indicating what this aims to address and maybe a high level of the approach?  Is this a bug fix? It'll very helpful for the reviewer but probably more so for future readers/debuggers/archeologists.

**Comment by Legogris** [comment]
Am I right in thinking this is a first step at addressing #8577 ? Is this expected to be included in 0.12.4 or is it a less short-term effort?

File: client/state/memdb.go [hunk 1]
```diff
@@ -89,6 +93,19 @@ func (m *MemDB) PutDeploymentStatus(allocID string, ds *structs.AllocDeploymentS
...
+func (m *MemDB) GetNetworkStatus(allocID string) (*structs.AllocNetworkStatus, error) {
+	m.mu.Lock()
+	defer m.mu.Unlock()
+	return m.networkStatus[allocID], nil
+}
+
+func (m *MemDB) PutNetworkStatus(allocID string, ns *structs.AllocNetworkStatus) error {
+	m.mu.Lock()
+	m.networkStatus[allocID] = ns
+	defer m.mu.Unlock()
```
Comment (shoenig) [inline client/state/memdb.go:105]: `defer` isn't needed if the unlock is after the critical section

**Review by shoenig — APPROVED** [review]
This looks great @nick!

**Review by Fernando938 — COMMENTED** [review]
[]()

**Review by Fernando938 — COMMENTED** [review]
[]()

**Pull request merged**